Allow UPnP to use a cached description URL

Yukitty opened the pull request:
UPnP discovery is slooow. But if you already know the description URL of the gateway, it can be near instant!

This PR adds miniupnpc's native `UPNP_GetIGDFromUrl` to Godot's UPNP API to allow for that possibility.

Simply save `upnp.get_gateway().description_url` to your game's globals or config file after a successful discovery, and next time your server needs to use UPnP, call `upnp.add_device_from_url(description_url)` instead of `discover`. If it fails (eg. because the user's network has changed), fallback to `discover` as normal. But if it succeeds, you can skip `discover` entirely and immediately start requesting ports!

Geometror commented:
Thanks for your first contribution :) 
Looks like the documentation is not up to date (since you bound a new function).
You need to call the godot executable with `--doctool .` in the projects root directory to regenerate the documentation and then edit the modified xml file manually to add some description of what your function does.

Yukitty commented:
Gosh, sorry for making so many little mistakes. 😥

Faless commented:
Seems to be leaking memory:
```
Direct leak of 40 byte(s) in 1 object(s) allocated from:
    #0 0x7f7d1f736808 in __interceptor_malloc ../../../../src/libsanitizer/asan/asan_malloc_linux.cc:144
    #1 0x56514b597771 in UPNP::add_device_from_url(String const&) modules/upnp/upnp.cpp:95
    #2 0x56514b5c8e06 in void call_with_variant_args_ret_helper<__UnexistingClass, int, String const&, 0ul>(__UnexistingClass*, int (__UnexistingClass::*)(String const&), Variant const**, Variant&, Callable::CallError&, IndexSequence<0ul>) core/variant/binder_common.h:680
    #3 0x56514b5c2d37 in void call_with_variant_args_ret_dv<__UnexistingClass, int, String const&>(__UnexistingClass*, int (__UnexistingClass::*)(String const&), Variant const**, int, Variant&, Callable::CallError&, Vector<Variant> const&) core/variant/binder_common.h:493
    #4 0x56514b5be46e in MethodBindTR<int, String const&>::call(Object*, Variant const**, int, Callable::CallError&) core/object/method_bind.h:465
    #5 0x56515dc94fb2 in Object::callp(StringName const&, Variant const**, int, Callable::CallError&) core/object/object.cpp:733
    #6 0x56515dc937b2 in Object::callv(StringName const&, Array const&) core/object/object.cpp:670
    #7 0x56515dd13dab in void call_with_variant_args_ret_helper<__UnexistingClass, Variant, StringName const&, Array const&, 0ul, 1ul>(__UnexistingClass*, Variant (__UnexistingClass::*)(StringName const&, Array const&), Variant const**, Variant&, Callable::CallError&, IndexSequence<0ul, 1ul>) core/variant/binder_common.h:680
    #8 0x56515dd0cf8e in void call_with_variant_args_ret_dv<__UnexistingClass, Variant, StringName const&, Array const&>(__UnexistingClass*, Variant (__UnexistingClass::*)(StringName const&, Array const&), Variant const**, int, Variant&, Callable::CallError&, Vector<Variant> const&) core/variant/binder_common.h:493
    #9 0x56515dd04d9c in MethodBindTR<Variant, StringName const&, Array const&>::call(Object*, Variant const**, int, Callable::CallError&) core/object/method_bind.h:465
    #10 0x565145f2b852 in GDScriptFunction::call(GDScriptInstance*, Variant const**, int, Callable::CallError&, GDScriptFunction::CallState*) modules/gdscript/gdscript_vm.cpp:1644
    #11 0x565145956985 in GDScriptInstance::callp(StringName const&, Variant const**, int, Callable::CallError&) modules/gdscript/gdscript.cpp:1634
    #12 0x56515dc94bb3 in Object::callp(StringName const&, Variant const**, int, Callable::CallError&) core/object/object.cpp:711
    #13 0x56515d24045a in Variant::callp(StringName const&, Variant const**, int, Variant&, Callable::CallError&) core/variant/variant_call.cpp:1048
    #14 0x565145f29383 in GDScriptFunction::call(GDScriptInstance*, Variant const**, int, Callable::CallError&, GDScriptFunction::CallState*) modules/gdscript/gdscript_vm.cpp:1555
    #15 0x565145956985 in GDScriptInstance::callp(StringName const&, Variant const**, int, Callable::CallError&) modules/gdscript/gdscript.cpp:1634
    #16 0x565152f7576d in bool Node::_gdvirtual__ready_call<false>() scene/main/node.h:242
    #17 0x565152ee2389 in Node::_notification(int) scene/main/node.cpp:154
    #18 0x565144708bdb in Node::_notificationv(int, bool) scene/main/node.h:45
    #19 0x56515dc964eb in Object::notification(int, bool) core/object/object.cpp:790
    #20 0x565152ee3c85 in Node::_propagate_ready() scene/main/node.cpp:187
    #21 0x565152f38dbe in Node::_set_tree(SceneTree*) scene/main/node.cpp:2533
    #22 0x565152f05aee in Node::_add_child_nocheck(Node*, StringName const&) scene/main/node.cpp:1117
    #23 0x565152f06fb6 in Node::add_child(Node*, bool, Node::InternalMode) scene/main/node.cpp:1136
    #24 0x565152ff62c5 in void call_with_variant_args_helper<__UnexistingClass, Node*, bool, Node::InternalMode, 0ul, 1ul, 2ul>(__UnexistingClass*, void (__UnexistingClass::*)(Node*, bool, Node::InternalMode), Variant const**, Callable::CallError&, IndexSequence<0ul, 1ul, 2ul>) core/variant/binder_common.h:262
    #25 0x565152fe6215 in void call_with_variant_args_dv<__UnexistingClass, Node*, bool, Node::InternalMode>(__UnexistingClass*, void (__UnexistingClass::*)(Node*, bool, Node::InternalMode), Variant const**, int, Callable::CallError&, Vector<Variant> const&) core/variant/binder_common.h:409
    #26 0x565152fd65b2 in MethodBindT<Node*, bool, Node::InternalMode>::call(Object*, Variant const**, int, Callable::CallError&) core/object/method_bind.h:320
    #27 0x565145f2b852 in GDScriptFunction::call(GDScriptInstance*, Variant const**, int, Callable::CallError&, GDScriptFunction::CallState*) modules/gdscript/gdscript_vm.cpp:1644
    #28 0x565145956985 in GDScriptInstance::callp(StringName const&, Variant const**, int, Callable::CallError&) modules/gdscript/gdscript.cpp:1634
    #29 0x565152f713db in bool Node::_gdvirtual__process_call<false>(double) scene/main/node.h:238

Direct leak of 4 byte(s) in 1 object(s) allocated from:
    #0 0x7f7d1f736c3e in __interceptor_realloc ../../../../src/libsanitizer/asan/asan_malloc_linux.cc:163
    #1 0x7f7d054eff3f in g_realloc (/lib/x86_64-linux-gnu/libglib-2.0.so.0+0x57f3f)
```

Likely due to the malloc without free. You can probably just use a stack variable alternatively.

Yukitty commented:
Ah... I was just blindly copying what I saw in `parse_igd` for that.
If just putting it on the stack is fine, mmm... That's honestly what it looks like how it was meant to be used anyway, tbh.

Yukitty commented:
😓 And for a moment I thought I was done embarassing myself...
Need to handle empty string correctly, I take it.

Faless commented:
> If just putting it on the stack is fine, mmm... That honestly looks like how it was meant to be used anyway, tbh.

Well, then the other function may be leaking too actually, `FreeUPNPUrls` does not free the struct pointer itself (as expected from a C API).

> And for a moment I thought I was done embarassing myself...

No worries, that's why we have CI tests, these small mistakes happens often even to seasoned contributors :)

Yukitty committed:
Allow skipping UPnP discovery by using cached IGD description URL

This is much faster than running a full discover,
but only works if the user's network hasn't changed.

Yukitty commented:
Squashed & rebased, nice and clean. 🐱

Yukitty commented:
Ah, looks like the CI just had a network hiccup and couldn't set up scons for Android...
That's fine, then.

AThousandShips commented on modules/upnp/upnp.h:
```suggestion
	UPNPResult add_device_from_url(const String &p_desc_url);
```
